### PR TITLE
Add performace curve structure to isentropic pressure changers.

### PR DESCRIPTION
**Pull request by eslickj**
Adding some support for performance curves in pressure changers.  Need these for gas turbines among other things.  This seems to work, but we can have some discussion around the design of it.  The main reason I used a sub-block for performance curves is that I didn't put any restriction on the form of the performance curves, so it gives me an easy way to deactivate and activate them.  The bad thing about it is that the variables and expressions that go into the performance cruves are defined in the parent block, so writing the performance curves is a little convoluted.  I can add some docs if everyone agrees to the implementation.  There are two ways to add curves, as seen in the tests.  Either a callback or just adding constraints after constructing the pressure changer.  Either way ends up the same. 

The main reason for adding this and not just slapping a few extra constraints into the existing turbine model is that I want to integrate with the pressure changer initialization, so people don't need to create some custom initialization if they want to use performance curves. 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: idaes/generic_models/unit_models/pressure_changer.py [hunk 1]
```diff
@@ -455,6 +479,38 @@ def actual_work(b, t):
                     b.work_isentropic[t] * b.efficiency_isentropic[t]
                 )
 
+        # Add head expressions (often calcuated from performance curve)
+        try:
+            @self.Expression(self.flowsheet().config.time)
+            def head_isentropic(b, t): # units are energy/mass e.g. J/kg
```
Comment (andrewlee94) [inline idaes/generic_models/unit_models/pressure_changer.py:485]: We should not be presuming set units here, as the PressureChanger gets it's units from the property package. We probably need to make use of unit conversions for this.

File: idaes/generic_models/unit_models/pressure_changer.py [hunk 2]
```diff
@@ -455,6 +479,38 @@ def actual_work(b, t):
                     b.work_isentropic[t] * b.efficiency_isentropic[t]
                 )
 
+        # Add head expressions (often calcuated from performance curve)
```
Comment (andrewlee94) [inline idaes/generic_models/unit_models/pressure_changer.py:482]: This comments doesn't seem to match what follows - it mentions performance curves, but they don't appear in the following `try` statement.

File: idaes/generic_models/unit_models/pressure_changer.py [hunk 3]
```diff
@@ -455,6 +479,38 @@ def actual_work(b, t):
...
+                        b.control_volume.properties_in[t].flow_mol /
+                        b.control_volume.properties_in[t].mw)
+
+        except PropertyNotSupportedError:
+            _log.warning(
+                "flow_mass or flow_mol and mw not supported by property package"
+                " but are required for isentropic pressure changer head"
+                " calculation")
+
+        # Add performace curve block.  If this block is empty that's fine
```
Comment (andrewlee94) [inline idaes/generic_models/unit_models/pressure_changer.py:510]: Should we put this in an `if` statement which checks to see if the user provides a performance curve object? If they don't we probably don't need to build this block.

**Review by andrewlee94 — COMMENTED** [review]
A few comments to start with.

File: idaes/generic_models/unit_models/pressure_changer.py (hunk 1 above)
Comment (eslickj) [reply idaes/generic_models/unit_models/pressure_changer.py:485]: I'm not assuming anything J/kg are just and example in the comment not necessarily the expected unit.  Whatever the property package uses is what it is.  I'll fix the comment to be clear.

**Review by TimBartholomew — COMMENTED** [review]
LGTM. I'll approve after Andrew's comments are addressed.

I appreciate the descriptive comment for the performance curve initialization, it would be nice if we had those for other initialization routines :)

File: idaes/generic_models/unit_models/pressure_changer.py (hunk 3 above)
Comment (eslickj) [reply idaes/generic_models/unit_models/pressure_changer.py:510]: It's set up so you can just add constraints to the performance_curve block later without the callback.  I thought the option may seem more straight forward to a lot of people, even-though it's not much different.

Comment (andrewlee94) [reply idaes/generic_models/unit_models/pressure_changer.py:510]: Also, just noted a typo: "performace"

**Comment by andrewlee94** [comment]
@eslickj I am still a bit unsure about this implementation, as it adds things to the basic isentropic pressure changer that are not needed in the simple cases - the main example being the properties needed for calculating head. Incidentally, this is what caused the test failures as the isentropic compressor was being used as a test for the properties interrogator tool, and the new properties showed up there which caused the tests to fail.

I think we need some way to only build the performance curve parts if the user requires them so that the simple case still works without needing the additioanl properties and without adding an unnecessary Block to the model.

File: idaes/generic_models/unit_models/pressure_changer.py [hunk 4]
```diff
@@ -479,34 +524,6 @@ def actual_work(b, t):
...
-
-        except PropertyNotSupportedError:
-            _log.warning(
-                "flow_mass or flow_mol and mw not supported by property package"
-                " but are required for isentropic pressure changer head"
-                " calculation")
-
         # Add performace curve block.  If this block is empty that's fine
         self.performance_curve = IsentropicPerformanceCurve(
             default=self.config.isentropic_perfomance_curves)
```
Comment (andrewlee94) [inline idaes/generic_models/unit_models/pressure_changer.py]: Typo: perfomance

This appears in a few places, so you might want to do a search and replace.

**Review by andrewlee94 — CHANGES_REQUESTED** [review]
Aside from a typo, I think this looks OK.

I am still not completely sold on always building the performance Block (mainly due to the memory overhead), but that isn't critical.

**Comment by eslickj** [comment]
@andrewlee94, I added an option so the performance curve block only gets added if you want it.  I guess I could simplify this a bit, but I'm still stuck on the idea of creating an empty block and letting the user just add curves to it later.  Otherwise I could probably just have a callback option, which if provided, creates and populates the performance curve block.

Maybe this would be simpler by callback only. Then instead of having the head expression option, I could just handle it by calling like ```build_head_expressions()``` in the callback.  Well now I'm not sure, but I guess it would be simpler to rework this and just require the callback.

**Comment by andrewlee94** [comment]
@eslickj My thought was maybe adding a configuration argument to the pressure changer model itself on whether to build the block or not. Creating an empty block so the user can fill it in later is fine, but I wonder if it would be better to let the user decide whether they want the block or not.

**Comment by eslickj** [comment]
@andrewlee94, I think I did that.  I must have snuck it past you in a recent commit.

**Comment by andrewlee94** [comment]
@eslickj Yes, you did sneak that change past me as I was reviewing the code - the new implementation looks good. All that is left is to document it.

**Comment by eslickj** [comment]
Alright docs are in.  Should be ready to go.

**Review by andrewlee94 — APPROVED** [review]
Looks OK to me.

**Review by andrewlee94 — DISMISSED** [review]
LGTM

**Comment by eslickj** [comment]
@andrewlee94 and @TimBartholomew, could you approve again.  The example code in the docs was failing its test just because it produced output.  It's fixed now.

**Review by andrewlee94 — APPROVED** [review]
Looks good.

**Review by TimBartholomew — APPROVED** [review]

**Pull request merged**